### PR TITLE
[4.2.0] Added createSourceWithParams method for Android (#112)

### PR DESCRIPTION
**Pull request by santiagofm**
- Updates Stripe to latest release
- Adds support for Bitcoin & AliPay payments in both platform Android & iOS

### CODE REVIEW

**Comment by isnifer** [comment]
Wow, good work. I will review this PR with my teammates tomorrow

File: README.md [hunk 1]
```diff
@@ -392,6 +392,191 @@ const token = await stripe.paymentRequestWithAndroidPay(options)
...
+
+Create a source to pay with `AliPay`. After the customer returns to the app, returns an object with the following data:
+* `sourceId` _String_ - A string representing the original ID of the `Source` object.
+* `clientSecret` _String_ - Used to confirm that the returning customer is the same one who triggered the creation of the source (source IDs are not considered secret)
+
+##### `options`
+An object with the following keys:
+
+* `amount` _Integer_ - A positive integer in the smallest currency unit representing the amount to charge the customer (e.g., 1099 for a $10.99 payment).
+* `currency` _String_ - The currency of the payment. Can be aud, cad, eur, gbp, hkd, jpy, nzd, sgd, or usd, defaults to usd.
```
Comment (paynd) [inline README.md]: Can be 'aud'|'cad'|'eur'|'gbp'|'hkd'|'jpy'|'nzd'|'sgd'|'usd, defaults to usd.

File: android/src/main/java/com/gettipsi/stripe/StripeModule.java [hunk 2]
```diff
@@ -243,6 +260,50 @@ public void paymentRequestWithAndroidPay(final ReadableMap map, final Promise pr
...
+        promise.resolve(convertSourceToWritableMap(source));
+      }
+    });
+  }
+
+  @ReactMethod
+  public void createSourceWithAliPay(final ReadableMap options, final Promise promise) {
+    SourceParams sourceParams = SourceParams.createCustomParams();
+    sourceParams.setType("alipay");
+    sourceParams.setCurrency(options.hasKey("currency") ? options.getString("currency") : "USD");
```
Comment (paynd) [inline android/src/main/java/com/gettipsi/stripe/StripeModule.java]: Should be "usd" lowercase?

File: android/src/main/java/com/gettipsi/stripe/StripeModule.java [hunk 3]
```diff
@@ -440,6 +502,123 @@ private WritableMap convertTokenToWritableMap(Token token) {
...
+    newSource.putMap("owner", convertOwnerToWritableMap(source.getOwner()));
+    newSource.putMap("receiver", convertReceiverToWritableMap(source.getReceiver()));
+    newSource.putMap("redirect", convertRedirectToWritableMap(source.getRedirect()));
+    newSource.putString("status", source.getStatus());
+    newSource.putString("type", source.getType());
+    newSource.putString("typeRaw", source.getTypeRaw());
+    newSource.putString("usage", source.getUsage());
+
+    // TODO: source.getCodeVerification(), it doesn't have public properties to get data.
+    // Waiting for PR https://github.com/stripe/stripe-android/pull/366 to be merged
```
Comment (paynd) [inline android/src/main/java/com/gettipsi/stripe/StripeModule.java]: Looks like it has been already merged.

File: android/src/main/java/com/gettipsi/stripe/StripeModule.java [hunk 4]
```diff
@@ -440,6 +502,123 @@ private WritableMap convertTokenToWritableMap(Token token) {
...
+    // newSource.putMap("codeVerification", source.getCodeVerification());
+
+    // TODO: source.getSourceTypeData(), it is a Map<String, Object> cannot be easily mapped to WritableMap
+    // newSource.putMap("sourceTypeData", source.getSourceTypeData());
```
Comment (paynd) [inline android/src/main/java/com/gettipsi/stripe/StripeModule.java]: I think better than nothing:
`newSource.putMap("sourceTypeData", mapToWritableMap(source.getSourceTypeData()));`

 ```java
 private WritableMap mapToWritableMap(Map<String, Object> map){
    WritableMap writableMap = Arguments.createMap();

    if (map == null) {
      return writableMap;
    }

    for(String key: map.keySet()){
      pushRightTypeToMap(writableMap, key, map.get(key));
    }

    return writableMap;
  }
```

```java
  private void pushRightTypeToMap(WritableMap map, String key, Object object) {
    Class argumentClass = object.getClass();
    if (argumentClass == Boolean.class) {
      map.putBoolean(key, (Boolean) object);
    } else if (argumentClass == Integer.class) {
      map.putDouble(key, ((Integer)object).doubleValue());
    } else if (argumentClass == Double.class) {
      map.putDouble(key, (Double) object);
    } else if (argumentClass == Float.class) {
      map.putDouble(key, ((Float)object).doubleValue());
    } else if (argumentClass == String.class) {
      map.putString(key, object.toString());
    } else if (argumentClass == WritableNativeMap.class) {
      map.putMap(key, (WritableNativeMap)object);
    } else if (argumentClass == WritableNativeArray.class) {
      map.putArray(key, (WritableNativeArray)object);
    } else {
      Log.e(TAG, "Can't map value of " + key + " to any valid js type,");
    }
  }
```

**Comment by santiagofm** [comment]
Thanks guys ! We'll fix that stuff & the merge conflicts in a bit.
Cheers

**Comment by isnifer** [comment]
@santiagofm we are planning to update test and merge iOS part first, next step your PR :)

**Comment by santiagofm** [comment]
Sounds good!

**Comment by isnifer** [comment]
So, @santiagofm @m-revetria I looked into your PR today and I want to say:
1. We don't need separate methods for bitcoin or Alipay, but we need common method `createSourceWithParams` like in https://github.com/tipsi/tipsi-stripe/pull/113. @flavordaaave did a good example in iOS how it looks like.
2. We have already the tests in his branch and we're planning to merge it into master today or tomorrow (depends on a free time of @cybergrind).
3. So, what you can do before rebase from `master`? Implement one method  `createSourceWithParams` instead of two for bitcoins and Alipay. Remove all iOS code, cause we have it already.
4. After rebase from `master` we can check e2e-tests for android too.

**Comment by jterps08** [comment]
Hi @isnifer @santiagofm It's exciting to see support for bitcoin here... Will this be merged soon?

**Comment by isnifer** [comment]
@jterps08 I hope. Guys don't respond, I need some time to do it.

**Comment by m-revetria** [comment]
@isnifer sorry for no answering before. We are still waiting for PR #113 to be merged before moving forward with the suggested changes.

Having that merged will make easier the work of updating this PR as we will have the base code for that. 

Any update about #113?

Regards

**Comment by isnifer** [comment]
@m-revetria ok, no problem. It ready to merge. And we will merge it ASAP

**Comment by isnifer** [comment]
@m-revetria @santiagofm So, guys. `master` branch has been updated. Please, rebase over master, remove your iOS implementation and do common method to create payments with Sources, not specific for Alipay or Bitcoins

**Comment by m-revetria** [comment]
@isnifer nice, thanks for letting us know, we will be working on that in these days!

**Comment by isnifer** [comment]
@m-revetria Awesome!

**Comment by m-revetria** [comment]
@isnifer we just rebased from master and applied the suggested changes.

Hope this gets merged soon! 😀

**Comment by isnifer** [comment]
@m-revetria Awesome! Will take a look tomorrow ASAP.

File: android/src/main/java/com/gettipsi/stripe/StripeModule.java [hunk 5]
```diff
@@ -456,6 +569,159 @@ private WritableMap convertTokenToWritableMap(Token token) {
...
+    newSource.putMap("receiver", convertReceiverToWritableMap(source.getReceiver()));
+    newSource.putMap("redirect", convertRedirectToWritableMap(source.getRedirect()));
+    newSource.putMap("sourceTypeData", mapToWritableMap(source.getSourceTypeData()));
+    newSource.putString("status", source.getStatus());
+    newSource.putString("type", source.getType());
+    newSource.putString("typeRaw", source.getTypeRaw());
+    newSource.putString("usage", source.getUsage());
+
+    // TODO: source.getCodeVerification(), it doesn't have public properties to get data.
+    // Waiting for a new stripe-android version containing PR https://github.com/stripe/stripe-android/pull/366
```
Comment (isnifer) [inline android/src/main/java/com/gettipsi/stripe/StripeModule.java]: Looks like it already there

File: android/src/main/java/com/gettipsi/stripe/StripeModule.java [hunk 6]
```diff
@@ -456,6 +569,159 @@ private WritableMap convertTokenToWritableMap(Token token) {
...
+    } else if (argumentClass == String.class) {
+      map.putString(key, object.toString());
+    } else if (argumentClass == WritableNativeMap.class) {
+      map.putMap(key, (WritableNativeMap)object);
+    } else if (argumentClass == WritableNativeArray.class) {
+      map.putArray(key, (WritableNativeArray) object);
+    } else {
+      Log.e(TAG, "Can't map "+ key + "value of " + argumentClass.getSimpleName() + " to any valid js type,");
+    }
+    // TODO: use https://github.com/facebook/facebook-android-sdk/blob/master/facebook/src/main/java/com/facebook/internal/BundleJSONConverter.java
```
Comment (isnifer) [inline android/src/main/java/com/gettipsi/stripe/StripeModule.java]: 404

File: example/android/app/build.gradle [hunk 7]
```diff
@@ -90,13 +90,13 @@ def enableSeparateBuildPerCPUArchitecture = false
 def enableProguardInReleaseBuilds = false
 
 android {
-    compileSdkVersion 25
-    buildToolsVersion "25.0.0"
+    compileSdkVersion 26
+    buildToolsVersion '26.0.1'
```
Comment (isnifer) [inline example/android/app/build.gradle]: Does Android developers have any guidelines about quotes?

File: src/Stripe.android.js [hunk 8]
```diff
@@ -53,10 +53,21 @@ class Stripe {
     checkInit(this)
     checkArgs(
       types.createTokenWithBankAccountParamsPropTypes,
-      params, 'params', 'Stripe.createTokenWithBankAccount'
+      params,
+      'params',
+      'Stripe.createTokenWithBankAccount'
```
Comment (isnifer) [inline src/Stripe.android.js]: what's wrong with it?

File: src/Stripe.android.js [hunk 9]
```diff
@@ -53,10 +53,21 @@ class Stripe {
...
     )
     return StripeModule.createTokenWithBankAccount(params)
-  }
+  };
```
Comment (isnifer) [inline src/Stripe.android.js]: no semi-colons, please

File: src/Stripe.android.js [hunk 10]
```diff
@@ -53,10 +53,21 @@ class Stripe {
...
+
+  createSourceWithParams = (params = {}) => {
+    checkInit(this)
+    checkArgs(
+      types.createSourceWithParamsPropType,
+      params, 'params', 'Stripe.createSourceWithParams'
+    )
+    return StripeModule.createSourceWithParams(params)
+  };
```
Comment (isnifer) [inline src/Stripe.android.js]: no semi-colons, please

**Comment by isnifer** [comment]
@m-revetria please do `git pull`, I have updated configs for Travis. Currently, Android app crashes on click on button `Create source for Alipay payment`

**Comment by isnifer** [comment]
@m-revetria you can check yourself locally in two steps:
1. Run android emulator (Android 5.1.1)
2. `npm run ci`

**Comment by m-revetria** [comment]
@isnifer Ok, looking into it

**Comment by isnifer** [comment]
@m-revetria just checked your progress. App really doesn't crash anymore, but it doesn't do anything after clicking on button in tests :) https://travis-ci.org/tipsi/tipsi-stripe/jobs/298701278
Hope, you are on the right way.

**Comment by m-revetria** [comment]
@isnifer I just finished with this, I had forgotten to handle the flow for new sources.

I added an activity `RedirectUriReceiver` to handle redirection after the user confirm payments. I added it in a way that simplifies the usage of this library. In the way we implemented here, developers doesn't have to add a new activity to get the redirection intent and forwarding it to tipsi-stripe. They just have to add some settings in their project's build.gradle file.

An alternative could be exposing a function in `StripeModule` to forward redirections. Developers will have to implement their own receiver activity and call that function. In my opinion, implementing this approach is a bit more complex for developers.

Please, let me know your thoughts about current implementation.

**Comment by isnifer** [comment]
@m-revetria I'm voting for the first option

**Comment by m-revetria** [comment]
@isnifer awesome! The first approach is already implemented in this PR. Can we move forward and merge it?

**Comment by isnifer** [comment]
@m-revetria only when tests will pass. Currently, when I click on the button to pay with `Sources`, I see loader and then nothing more. Please, check example app locally.

**Comment by m-revetria** [comment]
@isnifer weird, I solved that issue on commit [113c8ac](https://github.com/tipsi/tipsi-stripe/pull/115/commits/113c8ac481d2b79f27b417c4401b8b10e6446e3b).

Not sure why tests are failing, the output screenshot in travis shows the alipay source (at least for ios): https://i.imgur.com/e85Kb7G.png. I'm looking into this, if you have some advice on what could be the issue, it will be really helpful 😃

it's working fine for me, you should be able to see something like this:
alipay

**Comment by isnifer** [comment]
@m-revetria hmm, ok. I will run it in debug mode

**Comment by isnifer** [comment]
@m-revetria got this 
image

**Comment by isnifer** [comment]
@m-revetria ok, it works! I checked this on Android 7. Please, run it on Android 5.1.1. We support this version too

File: example/android/app/build.gradle (hunk 7 above)
Comment (paynd) [reply example/android/app/build.gradle]: Single quotes are a standard java String. Doublequotes - templattable string. But there is no special guidlines.

**Comment by isnifer** [comment]
@m-revetria Hi, there! Yesterday, @paynd said me that he added next line:
```java
browserIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
```

And got next result:
image

But, also, he got next error:
```
11-15 10:33:10.682 5847-5870/com.example E/StripeModule: Can't map statement_descriptorvalue of  to any valid js type,
11-15 10:33:10.682 5847-5870/com.example E/StripeModule: Can't map native_urlvalue of  to any valid js type,
11-15 10:33:10.682 5847-5870/com.example E/StripeModule: Can't map data_stringvalue of  to any valid js type,
```

**Comment by m-revetria** [comment]
Ok, I'm looking into it

**Comment by isnifer** [comment]
@m-revetria so, I talked with @paynd, he said that the second option you've described above:
```
An alternative could be exposing a function in StripeModule to forward redirections. Developers will have to implement their own receiver activity and call that function. In my opinion, implementing this approach is a bit more complex for developers.
```

Can help with Android <= 5

**Comment by paynd** [comment]
Hi m-revetria!
Unfortunately we have a problem with first implementation (you mentioned it [here]( https://github.com/tipsi/tipsi-stripe/pull/115#issuecomment-342926293)) on api below 24.
We still supports api from 22.
There are some implicit behaviour of android's activities back stack: I found that when we call `finish();` into the method `sendResult` of `RedirectUriReceiver` it should pops up user's react-native activity, but it doesen't.
I think we should use your second opportunity and expose a function in StripeModule to forward redirections.

**Comment by isnifer** [comment]
@m-revetria please, rebase over `master` branch

**Comment by m-revetria** [comment]
@isnifer @paynd the issue on Android 5 is fixed on last commit. I'm still using the following approach: the redirect is handled in tipsi-stripe, no function is being exposed to developers. 

Additionally, the last commit manages the case where the user closes the browser before completing the payment flow.

**Comment by isnifer** [comment]
@m-revetria awesome, I will wait for tests 👍

**Comment by isnifer** [comment]
@m-revetria awesome, it fails on the browser. I will update tests

**Comment by isnifer** [comment]
Looks like there is a problem with a click in the old browser for Android. Appium clicks, but nothing happened. I've disabled tests for Sources for Android until we'll find a solution.
I've checked this PR manually on Android 5 and it works.

@m-revetria @santiagofm awesome work, thank you guys for your contribution!

**Comment by isnifer** [comment]
Finally, I did it.

**Comment by cybergrind** [comment]
Thanks to all. @m-revetria and @isnifer good job

**Pull request merged**